tests/unit/test_download.py

```diff
@@ -157,6 +158,74 @@ def test_unpack_http_url_bad_downloaded_checksum(mock_unpack_file):
         rmtree(download_dir)
 
 
+def test_download_http_url_with_directory_traversal(tmpdir):
+    """
+    Test downloading a file when the content-disposition header
+    contains a filename with a ".." path part.
+    It should download file to the tmp_subdir.
+    """
+    mock_url = 'http://www.example.com/whatever.tgz'
+    contents = b'downloaded'
+    link = Link(mock_url)
+
+    session = Mock()
+    resp = MockResponse(contents)
+    resp.url = mock_url
+    resp.headers = {
+        # content-type set random avoid mimetypes.guess_extension guess ext
+        'content-type': 'random',
+        'content-disposition': 'attachment;filename="../out_dir_file"'
+    }
+    session.get.return_value = resp
+
+    tmp_subdir = mkdtemp(dir=tmpdir, prefix="sub-dir-")
+    file_path, content_type = _download_http_url(
+        link,
+        session,
+        tmp_subdir,
+        hashes=None,
+        progress_bar="on"
+    )
+    assert set(os.listdir(tmp_subdir)) == {
+        'out_dir_file'
+    }
+
+
+@pytest.mark.parametrize(
+    "content_disposition, default_filename, win_expected, non_win_expected",
+    [
+        ('', 'df', 'df', 'df'),
+        ('random', 'df', 'df', 'df'),
+        ('attachment;filename="dir/file"', 'df', 'file', 'file'),
+        ('attachment;filename="../file"', 'df', 'file', 'file'),
+        ('attachment;filename="../../file"', 'df', 'file', 'file'),
+        ('attachment;filename="../"', 'df', 'df', 'df'),
+        ('attachment;filename="../.."', 'df', '..', '..'),
+        ('attachment;filename="/"', 'df', 'df', 'df'),
+        ('attachment;filename="dir\\file"', 'df', 'file', 'dir\\file'),
+        ('attachment;filename="..\\file"', 'df', 'file', '..\\file'),
+        ('attachment;filename="..\\..\\file"', 'df', 'file', '..\\..\\file'),
+        ('attachment;filename="..\\"', 'df', 'df', '..\\'),
+        ('attachment;filename="..\\.."', 'df', '..', '..\\..'),
+        ('attachment;filename="\\"', 'df', 'df', '\\'),
+    ]
+)
+def test_parse_content_disposition(
```

Comment:
I think it will be clearer if you divide this unit test into two tests: one for the test cases where the two test expectations are the same, and one for where they're different.